File: src/vibe_kanban_clone/services/runs.py
"""Task run execution service."""

from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from vibe_kanban_clone.api.routes.ws import broadcast_project
from vibe_kanban_clone.executors.registry import get_executor
from vibe_kanban_clone.models.task import Task
from vibe_kanban_clone.models.task_run import TaskRun
from vibe_kanban_clone.services.models import get_model


async def list_task_runs(session: AsyncSession, task_id: int) -> list[TaskRun]:
    result = await session.scalars(
        select(TaskRun).where(TaskRun.task_id == task_id).order_by(TaskRun.started_at.desc())
    )
    return list(result.all())


async def execute_task_run(
    async_session_factory: async_sessionmaker,
    task_id: int,
    model_id: int,
    prompt: str,
) -> TaskRun:
    """Execute a task run with streaming and broadcast updates."""
    async with async_session_factory() as session:
        task = await session.get(Task, task_id)
        if task is None:
            raise RuntimeError("Task not found")

        run = TaskRun(
            task_id=task_id,
            model_id=model_id,
            agent_id=task.assigned_agent_id,
            status="running",
            prompt=prompt,
            started_at=datetime.now(UTC),
        )
        session.add(run)
        await session.commit()
        await session.refresh(run)

        await broadcast_project(
            task.project_id,
            "run.started",
            {"run_id": run.id, "task_id": task.id},
        )

        model = await get_model(session, model_id)
        if model is None:
            raise RuntimeError("Model not found")

        executor = get_executor(model.provider)

    try:
        full_output = ""
        async for token in executor.run(
            prompt,
            {
                "model_id": model.model_id,
                "api_key_env": model.api_key_env,
                "base_url": model.base_url,
            },
        ):
            full_output += token
            await broadcast_project(
                task.project_id,
                "run.token",
                {"run_id": run.id, "token": token},
            )

        async with async_session_factory() as session:
            run = await session.get(TaskRun, run.id)
            if run is not None:
                run.output = full_output
                run.status = "completed"
                run.finished_at = datetime.now(UTC)
                await session.commit()

        await broadcast_project(
            task.project_id,
            "run.finished",
            {"run_id": run.id, "status": "completed"},
        )
        return run
    except Exception as e:
        async with async_session_factory() as session:
            run_clone = await session.get(TaskRun, run.id)
            if run_clone is not None:
                run_clone.status = "failed"
                run_clone.error = f"{type(e).__name__}: {e}"[:500]
                run_clone.finished_at = datetime.now(UTC)
                await session.commit()

        await broadcast_project(
            task.project_id,
            "run.finished",
            {"run_id": run.id, "status": "failed", "error": f"{type(e).__name__}: {e}"[:500]},
        )
        return run_clone

File: server/src/services/runs.ts
import { prisma } from "../prisma/client.js";
import { logger } from "../lib/logger.js";
import { getExecutor } from "../executors/registry.js";
import { getModel } from "./models.js";
import { broadcastProject } from "../ws/manager.js";
import type { ExecutorContext } from "../executors/base.js";

const RUN_TIMEOUT_MS = 300_000;

export async function listTaskRuns(taskId: number, limit = 50, offset = 0) {
  return prisma.taskRun.findMany({
    where: { task_id: taskId },
    orderBy: { started_at: "desc" },
    take: limit,
    skip: offset,
  });
}

export async function executeTaskRun(
  taskId: number,
  modelId: number,
  prompt: string
) {
  const task = await prisma.task.findUnique({ where: { id: taskId } });
  if (!task) throw new Error("Task not found");

  const run = await prisma.taskRun.create({
    data: {
      task_id: taskId,
      model_id: modelId,
      agent_id: task.assigned_agent_id,
      status: "running",
      prompt,
      started_at: new Date(),
    },
  });

  await broadcastProject(task.project_id, "run.started", {
    run_id: run.id,
    task_id: task.id,
  });

  const model = await getModel(modelId);
  if (!model) throw new Error("Model not found");

  const executor = getExecutor(model.provider);

  let context: ExecutorContext | undefined;
  if (task.assigned_agent_id) {
    const agent = await prisma.agent.findUnique({
      where: { id: task.assigned_agent_id },
      include: { agentSkills: { include: { skill: true } } },
    });
    if (agent) {
      context = {
        systemPrompt: agent.system_prompt ?? undefined,
        skills: agent.agentSkills.map((as) => as.skill),
      };
    }
  }

  let output = "";
  try {
    const iterable = executor.run(prompt, {
      model_id: model.model_id,
      api_key_env: model.api_key_env,
      base_url: model.base_url,
    }, context);

    const iterator = iterable[Symbol.asyncIterator]();
    const startTime = Date.now();

    while (true) {
      const remaining = RUN_TIMEOUT_MS - (Date.now() - startTime);
      if (remaining <= 0) throw new Error("Task run timed out");

      const { value, done } = await Promise.race([
        iterator.next(),
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error("Task run timed out")), remaining)
        ),
      ]);

      if (done) break;
      output += value;
      await broadcastProject(task.project_id, "run.token", {
        run_id: run.id,
        token: value,
      });
    }

    await prisma.taskRun.update({
      where: { id: run.id },
      data: { status: "completed", output, finished_at: new Date() },
    });

    await broadcastProject(task.project_id, "run.finished", {
      run_id: run.id,
      status: "completed",
    });

    return run;
  } catch (e) {
    const error = `${e instanceof Error ? e.constructor.name : "Error"}: ${e instanceof Error ? e.message : String(e)}`.slice(0, 500);

    await prisma.taskRun.update({
      where: { id: run.id },
      data: { status: "failed", error, finished_at: new Date() },
    });

    await broadcastProject(task.project_id, "run.finished", {
      run_id: run.id,
      status: "failed",
      error,
    });

    logger.error({ run_id: run.id, error }, "task_run_failed");
    return run;
  }
}
